source.go

```diff
@@ -193,13 +194,33 @@ func (s *Source) AsMount(name string, sOpt SourceOpts, opts ...llb.ConstraintsOp
 	return s.asState(name, true, sOpt, opts...)
 }
 
+var errInvalidMountConfig = errors.New("invalid mount config")
+
+func (m *SourceMount) validate(root string) error {
+	if m.Dest == "/" {
+		return errors.Wrap(errInvalidMountConfig, "mount destination must not be \"/\"")
+	}
+	if root != "/" && filepath.HasPrefix(m.Dest, root) {
```

Comment:
Checking the output of `filepath.Rel` is a possible alternative?

Reply:
`filepath.Rel` doesn't really work here because it will always give you the combined path regardless of if they are related.

I'ave added a more thorough implementation that deals with a few different things.